src/Microsoft.AspNetCore.OData/EnableQueryAttribute.cs

```diff
@@ -38,6 +41,177 @@ public partial class EnableQueryAttribute : ActionFilterAttribute
+                    return;
+                }
+
+                Type clrType = edmModel.GetTypeMappingCache().GetClrType(
+                    elementType.ToEdmTypeReference(false),
+                    edmModel);
+
+                // CLRType can be missing if untyped registrations were made.
+                if (clrType != null)
+                {
+                    queryContext = new ODataQueryContext(edmModel, clrType, path);
+                }
+                else
+                {
+                    // In case where CLRType is missing, $count, $expand verifications cannot be done.
+                    // More importantly $expand required ODataQueryContext with clrType which cannot be done
+                    // If the model is untyped. Hence for such cases, letting the validation run post action.
+                    _queryValidationRunBeforeActionExecution = false;
+                    return;
+                }
+
+                _queryValidationRunBeforeActionExecution = true;
+            }
+            else
+            {
+                // For non-OData Json based controllers.
+                // For these cases few options are supported like IEnumerable<T>, Task<IEnumerable<T>>, T, Task<T>
+                // Other cases where we cannot determine the return type upfront, are not supported
+                // Like IActionResult, SingleResult. For such cases, the validation is run in OnActionExecuted
+                // When we have the result.
+                ControllerActionDescriptor controllerActionDescriptor = context.ActionDescriptor as ControllerActionDescriptor;
+
+                Type returnType = controllerActionDescriptor.MethodInfo.ReturnType;
+                Type elementType;
+
+                // For Task<> get the base object.
+                if (returnType.IsGenericType && returnType.GetGenericTypeDefinition() == typeof(Task<>))
+                {
+                    returnType = returnType.GetGenericArguments().First();
+                }
+
+                // For NetCore2.2+ new type ActionResult<> was created which encapculates IActionResult and T result.
+                // However we don't exactly have a version specific to NetCore2.2 (also at the time of writing this code
+                // 2.2 and 3.0 are both out of support), hence the code is made to work on NetCore3.1+ only.
+#if NETCOREAPP3_1 || NET5_0
+                if (returnType.IsGenericType && returnType.GetGenericTypeDefinition() == typeof(ActionResult<>))
+                {
+                    returnType = returnType.GetGenericArguments().First();
+                }
+#endif
+                if (TypeHelper.IsCollection(returnType))
+                {
+                    elementType = TypeHelper.GetImplementedIEnumerableType(returnType);
+                }
+                else if(TypeHelper.IsGenericType(returnType) && returnType.GetGenericTypeDefinition() == typeof(Task<>))
```

Comment:
Wondering if this block and [this other block](https://github.com/OData/WebApi/pull/2362/files#diff-1cea7432b6ec6bf7cc6472857ecbea4b4d6303185cc8f6cd6aa00d6405fd5985R128) can't be combined

Reply:
Sorry it is not sure from the comment which other block you mean. If you mean line 137, then that is specialized handling for a case only present NetCore31 and above.

Reply:
I meant Line 135 that is not within the `#if NETCOREAPP3_1 || NET5_0` block
![image](https://user-images.githubusercontent.com/28291332/101454407-00dc6280-3942-11eb-8e02-ab93e3df86b1.png)

Reply:
ActionResult<> has Task<> and other result types inside it, so it can't be merged. It is a boxing type in that sense.